### PR TITLE
LdapUtils refactor

### PR DESCRIPTION
**Pull request by rvazarkar**
# This MR is not intended to be merged. This is for review purposes

## Description

Completely refactors connection logic and the API contract. 

## Motivation and Context

LdapUtils was a complicated spaghetti mess. This allowed us to fix our API contract, make the library completely async/await supported and clean up our connection logic significantly for compatibility along with fixing several outstanding bugs and implementing new features

## Types of changes

-   [ ] Chore (a change that does not modify the application functionality)
-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

-   [ ] Documentation updates are needed, and have been made accordingly.
-   [x] I have added and/or updated tests to cover my changes.
-   [x] All new and existing tests passed.
-   [ ] My changes include a database migration.

### CODE REVIEW

File: src/CommonLib/Exceptions/LdapAuthenticationException.cs [hunk 1]
```diff
@@ -5,8 +5,10 @@ namespace SharpHoundCommonLib.Exceptions
 {
     public class LdapAuthenticationException : Exception
```
Comment (rvazarkar) [inline src/CommonLib/Exceptions/LdapAuthenticationException.cs]: Make these exceptions private

File: src/CommonLib/LdapConnectionPool.cs [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+            string[] attributes) {
+            var searchRequest = new SearchRequest(distinguishedName, ldapFilter,
+                searchScope, attributes);
+            searchRequest.Controls.Add(new SearchOptionsControl(SearchOption.DomainScope));
+            return searchRequest;
+        }
+    }
+}
+
+//TESTLAB
```
Comment (rvazarkar) [inline src/CommonLib/LdapConnectionPool.cs]: Delete these comments

File: test/unit/CommonLibTest.csproj [hunk 3]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
     <PropertyGroup>
-        <TargetFramework>net5.0</TargetFramework>
```
Comment (urangel) [inline test/unit/CommonLibTest.csproj:4]: Suggest updating the dockerfile to also target 7.0

File: test/unit/ComputerSessionProcessorTest.cs [hunk 4]
```diff
@@ -34,7 +34,7 @@ public void Dispose()
 
         #endregion
 
-        [WindowsOnlyFact]
```
Comment (urangel) [inline test/unit/ComputerSessionProcessorTest.cs:37]: nice updates 📠

File: test/unit/DomainTrustProcessorTest.cs [hunk 5]
```diff
@@ -53,52 +54,50 @@ public void DomainTrustProcessor_EnumerateDomainTrusts_HappyPath()
...
                     new Dictionary<string, object>
                     {
                         {"trustdirection", "3"},
                         {"trusttype", "2"},
                         {"trustattributes", 0x24.ToString()},
                         {"cn", "external.local"},
                         {"securityIdentifier", Array.Empty<byte>()}
-                    }, "", Label.Domain),
-                new MockSearchResultEntry("CN\u003dexternal.local,CN\u003dSystem,DC\u003dtestlab,DC\u003dlocal",
+                    }, "","")),
```
Comment (urangel) [inline test/unit/DomainTrustProcessorTest.cs:70]: Is it possible to set default values of empty strings for these last two params in MockDirectoryObject?

**Review by urangel — APPROVED** [review]
Left some comments but they are not critical

File: test/unit/DomainTrustProcessorTest.cs (hunk 5 above)
Comment (rvazarkar) [reply test/unit/DomainTrustProcessorTest.cs:70]: Absolutely

**Review by juggernot325 — APPROVED** [review]

**Pull request closed without merging**